### PR TITLE
add casync support to agnos updater

### PR DESCRIPTION
**Pull request by pd0wm**
Before using:
- [ ] Write tests for low level casync routines
- [ ] Add previous agnos images to azure
- [ ] Add scripts to push new agnos images as casync

```
casync make --compression=xz --store system-11fdbc9e8a9cd27f98346d7e1039bc5b3032d0e892ff95fa1258673ff1809bca system-11fdbc9e8a9cd27f98346d7e1039bc5b3032d0e892ff95fa1258673ff1809bca.caibx system-11fdbc9e8a9cd27f98346d7e1039bc5b3032d0e892ff95fa1258673ff1809bca.img.raw
```

agnos 5.1 -> 5.2 with a zeroed out target partition:

```
ERROR:root:casync done {"remote": 925660458, "seed": 9811655471, "target": 102311}
uncompressed stats:
 - remote: 882 MB
 - seed: 9357 MB

actual download (compressed): 175MB
```

### CODE REVIEW

File: system/hardware/tici/agnos.py [hunk 1]
```diff
@@ -172,9 +165,68 @@ def flash_partition(target_slot_number: int, partition: dict, cloudlog):
...
+
+  def progress(cur):
+    nonlocal last_p
+    p = int(cur / partition['size'] * 100)
+    if p != last_p:
+      last_p = p
+      print(f"Installing {partition['name']}: {p}", flush=True)
+
+  stats = casync.extract(target, sources, path, progress)
+  cloudlog.error(f'casync done {json.dumps(stats)}')
```
Comment (adeebshihadeh) [inline system/hardware/tici/agnos.py:200]:
```suggestion
  cloudlog.event(f'casync done', stats=stats, error=True)
```

File: system/hardware/tici/agnos.py [hunk 2]
```diff
@@ -172,9 +165,68 @@ def flash_partition(target_slot_number: int, partition: dict, cloudlog):
...
+  seed_path = path[:-1] + ('b' if path[-1] == 'a' else 'a')
+
+  target = casync.parse_caibx(partition['casync_caibx'])
+
+  sources = []
+
+  # First source is the current partition. Index file for current version is provided in the manifest
+  # TODO: build url based on agnos version or parition hash instead?
+  if 'casync_seed_caibx' in partition:
+    sources += [('seed', functools.partial(casync.read_chunk_local_file, f=open(seed_path, 'rb')), casync.build_chunk_dict(casync.parse_caibx(partition['casync_seed_caibx'])))]
```
Comment (adeebshihadeh) [inline system/hardware/tici/agnos.py:182]: we should be better about using context manager or explicitly closing files. this is going to be a problem once we start setting PYTHONWARNINGS=error in CI

File: system/hardware/tici/agnos.py [hunk 3]
```diff
@@ -172,9 +165,68 @@ def flash_partition(target_slot_number: int, partition: dict, cloudlog):
...
+
+  # Second source is the target partition, this allows for resuming
+  sources += [('target', functools.partial(casync.read_chunk_local_file, f=open(path, 'rb')), casync.build_chunk_dict(target))]
```
Comment (adeebshihadeh) [inline system/hardware/tici/agnos.py:185]: same here

File: system/hardware/tici/agnos.py [hunk 4]
```diff
@@ -172,9 +165,68 @@ def flash_partition(target_slot_number: int, partition: dict, cloudlog):
...
+  last_p = 0
+
+  def progress(cur):
+    nonlocal last_p
+    p = int(cur / partition['size'] * 100)
+    if p != last_p:
+      last_p = p
+      print(f"Installing {partition['name']}: {p}", flush=True)
+
+  stats = casync.extract(target, sources, path, progress)
```
Comment (adeebshihadeh) [inline system/hardware/tici/agnos.py:199]: would especially add types for the sources

**Review by adeebshihadeh — APPROVED** [review]

File: system/hardware/tici/agnos.py (hunk 1 above)
Comment (pd0wm) [reply system/hardware/tici/agnos.py:199]: We pass in a normal logger in standalone mode, which doesn't have `event`

**Pull request merged**